fix file attributes changed detection

tchernomax opened the pull request:
##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

module_utils/basic

##### ANSIBLE VERSION

```
ansible 2.3.0
```

##### SUMMARY

https://docs.python.org/2/library/stdtypes.html#str.split
str.split([sep[, maxsplit]])
If sep is given, consecutive delimiters are not grouped together and are deemed
to delimit empty strings.

>>> "85563      ----------------C-- /var/lib/libvirt/images".split(' ')[0:2]
['85563', '']
>>> "85563      ----------------C-- /var/lib/libvirt/images".split()[0:2]
['85563', '----------------C--']

Commit:
fix file attributes changed detection

https://docs.python.org/2/library/stdtypes.html#str.split
str.split([sep[, maxsplit]])
If sep is given, consecutive delimiters are not grouped together and are deemed
to delimit empty strings.

>>> "85563      ----------------C-- /var/lib/libvirt/images".split(' ')[0:2]
['85563', '']
>>> "85563      ----------------C-- /var/lib/libvirt/images".split()[0:2]
['85563', '----------------C--']

calfonso commented:
@tchernomax would you mind detailing the need to splitting on non-spaces?

calfonso commented:
needs_info

jctanner commented:
@tchernomax can you please revise this to do the split and then check the length before using indexes that may or may not exist?

needs_revision

tchernomax commented:
@jctanner done
@calfonso sometimes `lsattr -vd …` doesn't separate elements by just one space but by several spaces.
```
max@max-laptop % lsattr -vd /var/lib/libvirt/images
85563      ----------------C-- /var/lib/libvirt/images
```
As the python documentation says : If sep is given, consecutive delimiters are not grouped together and are deemed.
```python
>>> "85563      ----------------C-- /var/lib/libvirt/images".split(' ')
['85563', '', '', '', '', '', '----------------C--', '/var/lib/libvirt/images']
>>> "85563      ----------------C-- /var/lib/libvirt/images".split(' ')[0:2]
['85563', '']
```
We obviously want `['85563', '----------------C--']` not `['85563', '']`.
```python
>>> "85563      ----------------C-- /var/lib/libvirt/images".split()
['85563', '----------------C--', '/var/lib/libvirt/images']
>>> "85563      ----------------C-- /var/lib/libvirt/images".split()[0:2]
['85563', '----------------C--']
```

gundalow commented:
Rebase needed

tchernomax commented:
@gundalow done

tchernomax commented:
Any update on this ?

tchernomax commented:
another rebase…

tchernomax commented:
@gundalow re-re-rebase

tchernomax commented:
##### ANSIBLE VERSION

```
ansible 2.5.0 (vmware_host 7fe1b45ba6) last updated 2017/09/19 23:52:13 (GMT +200)
  config file = /home/max/.ansible.cfg
  configured module search path = ['/home/max/.ansible/modules']
  ansible python module location = /home/max/Documents/repositories-github/ansible/lib/ansible
  executable location = /home/max/Documents/repositories-github/ansible/bin/ansible
  python version = 3.6.2 (default, Jul 20 2017, 03:52:27) [GCC 7.1.1 20170630]
```

pilou- commented on test/integration/targets/file/tasks/main.yml:
```diff
@@ -62,6 +62,22 @@
+- name: verify that the file was not marked as changed
+  assert:
+    that:
+      - "file_attributes_result_2.changed == false"
```
Use test instead of filter, replace both `.changed` and `|changed` with `is changed` ([see](https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/porting_guide_2.5.rst#jinja-tests-used-as-filters)).

tchernomax replied:
done

pilou- commented on test/integration/targets/file/tasks/main.yml:
```diff
@@ -62,6 +62,22 @@
+  file: path={{output_dir}}/baz.txt attributes=A
+  register: file_attributes_result
+  ignore_errors: True
+
```
Should not an assertion be added there (then `when:` attribute of tasks below could be removed) ?
```
assert:
  that:
    file_attributes_result|changed
```

tchernomax replied:
No, the file module is [designed to fail](https://github.com/tchernomax/ansible/blob/devel/lib/ansible/module_utils/basic.py#L1282) if the extended attribute can't be applied.
On the contrary the get_file_attributes is [designed not to fail](https://github.com/tchernomax/ansible/blob/devel/lib/ansible/module_utils/basic.py#L1299) if `lsattr` isn't present or plate-forme doesn't support extended attributes.

If I where to replate `ignore_errors: True` with what you propose, the test would fail on all the plate-forme which doesn't support `chattr` or the 'A' extended attributes.

What I want to test is: **if** the plate-forme support the 'A' extended attribute **and** this attribute is already set ; the file module should **not** change the file.

pilou- reviewed (changes requested):
Updated integration tests don't fail when using devel version of `lib/ansible/module_utils/basic.py`. I tested the following commands on *devel* branch:
1. wget -O test/integration/targets/file/tasks/main.yml https://raw.githubusercontent.com/tchernomax/ansible/d759f30ec8215450a79969dedae740baf558da13/test/integration/targets/file/tasks/main.yml
2. ansible-test integration --docker ubuntu1604py3 --docker-privileged file

The integration tests must check that the fix is effective: integration tests can not be successful without the fix.

tchernomax commented:
@pilou- 

> Updated integration tests don't fail when using devel version of lib/ansible/module_utils/basic.py. …

That just mean your file system doesn't support chattr or the 'A' extended attribute, which make my patch impossible to test on your plateform ⇒ the integration test don't fail.

On my plate-form (Archlinux with btrfs):
```
% git clone https://github.com/ansible/ansible.git --recursive
% cd ansible
% wget -O test/integration/targets/file/tasks/main.yml https://raw.githubusercontent.com/tchernomax/ansible/c81a5f1e51dc171933bc8ce3f1394895943cca1b/test/integration/targets/file/tasks/main.yml
% source ./hacking/env-setup
% sudo ansible-test integration file
…
TASK [file : change file attribute "A"] ******************************************************************
changed: [testhost]

TASK [file : reapply file attribute "A"] *****************************************************************
changed: [testhost]

TASK [file : verify that the file was not marked as changed] *********************************************
fatal: [testhost]: FAILED! => {
    "assertion": "file_attributes_result_2 is not changed",
    "changed": false,
    "evaluated_to": false
}
        to retry, use: --limit @/home/max/Telechargements/ansible/test/integration/file-5i78sdfe.retry

PLAY RECAP ***********************************************************************************************
testhost                   : ok=13   changed=5    unreachable=0    failed=1
```

tchernomax commented:
@pilou- @jctanner as I mentioned on IRC, I did [some research](https://github.com/ansible/ansible/pull/33870) and it seems either lsattr/chattr [isn't installed](https://app.shippable.com/github/ansible/ansible/runs/47294/27/console) either [extended attributes aren't supported](https://app.shippable.com/github/ansible/ansible/runs/47294/32/console).
In fact, [extended attributes aren't supported on AUFS](https://github.com/moby/moby/issues/1070#issuecomment-20624664), which is used by the github ansible docker test suites (shippable). So even if I were to install the lsattr/chattr package when it's missing… those commands would fail.

What I propose is to keep the integration test I wrote, as it "pass" when extended attributes are supported, and "skip" when they aren't. And merge this PR as it is.
(I just did a rebase)

Please be cautious if you ask anything more… It's been A YEAR since I first wrote this, basically, two lines PR. I am beginning to feel discouraged.

tchernomax commented:
A precision though, as @mattclay tels me, RHEL integrations test are run on full VM ; so they support extended attribute.
The test on https://github.com/ansible/ansible/pull/33870 ; https://app.shippable.com/github/ansible/ansible/runs/47294/6/console didn't failed because RHEL use an old e2fsprogs version.

The [lsattr output changed](https://git.kernel.org/pub/scm/fs/ext2/e2fsprogs.git/commit/?h=debian-packaging&id=1e3f5c889d3f86f88dca94cd93dcfbe9aa37d641) in may 2016, with version 1.43:
```
max@mde-oxalide % git log 1e3f5c889d3f86f88dca94cd93dcfbe9aa37d641~1..1e3f5c889d3f86f88dca94cd93dcfbe9aa37d641
commit 1e3f5c889d3f86f88dca94cd93dcfbe9aa37d641
Author: Andreas Dilger <adilger@dilger.ca>
Date:   Sat Mar 5 17:38:34 2016 -0700

    lsattr: treat inode generation as an unsigned int

    The EXT2_GETVERSION ioctl is defined to take a "long" parameter, but
    fgetversion() calls ioctl() with an "int" parameter instead.  This is
    handled in the kernel correctly, but the generation is sign-extended
    in fgetversion() before return on 64-bit systems and lsattr prints
    it as a huge positive number for inode generation above 0x80000000:

            1635574212 -------------e-- /mnt/ost0/O/0/d0/12928
            18446744073045131735 -------------e-- /mnt/ost0/O/0/d0/166240
            782808861 -------------e-- /mnt/ost0/O/0/d0/31744
            18446744072181134840 -------------e-- /mnt/ost0/O/0/d0/135008

    Correctly assign the returned generation number as an unsigned value,
    and print it with a 10-character field width.  The version is printed
    left-aligned for consistency with the old code and to ensure it is
    always printed in the first column for use with tools like "cut":

            1635574212 -------------e-- /mnt/ost0/O/0/d0/12928
            3630547415 -------------e-- /mnt/ost0/O/0/d0/166240
            782808861  -------------e-- /mnt/ost0/O/0/d0/31744
            2766550520 -------------e-- /mnt/ost0/O/0/d0/135008

    Do not return a random value from the stack as the version on error.
    Clean up some style issues and consolidate some duplicate code.

    Signed-off-by: Andreas Dilger <adilger@dilger.ca>
    Signed-off-by: Theodore Ts'o <tytso@mit.edu>

max@mde-oxalide % git tag --contains=1e3f5c889d3f86f88dca94cd93dcfbe9aa37d641
debian-1.43-2
debian-1.43.2-2
debian-1.43.3-1
v1.43
v1.43-WIP-2016-03-15
v1.43-WIP-2016-05-12
v1.43.1
v1.43.2
v1.43.3
v1.43.4
v1.43.5
v1.43.6
v1.43.7

max@mde-oxalide % git log v1.43~1..v1.43
commit d6adf070b0e85f209c0d7f310188b134b5cb7180 (tag: v1.43)
Author: Theodore Ts'o <tytso@mit.edu>
Date:   Tue May 17 01:14:25 2016 -0400

    Update release notes for a 1.43 (17-Mar-2016)
```

On version 1.43 and more the file attributes can be separated from the version/generation number by more than one space.

So this PR fix a bug which only appear on plate-form which use e2fsprogs 1.43 or later.

pilou- commented:
Just in case, here is some code testing e2fsprogs version: https://github.com/ansible/ansible/pull/25519/files#diff-70ff37ad6549f1042f435c9cf8b63626R34.

tchernomax commented:
rebased

abadger commented:
Fixes #33745

abadger commented:
Talked this over on IRC and agreed with https://github.com/ansible/ansible/pull/18731#issuecomment-351497996

Merged to devel and cherry-picked to stable-2.4 for 2.4.3 beta2

abadger commented:
@tchernomax thanks for your work on this and sorry it took so long to get it merged!
